### PR TITLE
Completely overhaul grid utils + `min_Δx/y/z` -> `minimum_spacing` + move `x/y/zspacing` to `Grids`

### LINKED ISSUE #1119: Mixing data types and instantiated types in the user interface
More of a general issue that #1118 made me think about is that we've been mixing data types and instantiated types.

For example, we use data types when constructing grid topologies and specifying locations for abstract operations:

```julia
topology = (Periodic, Bounded, Bounded)
kinetic_energy @at (Cell, Cell, Cell) (u^2 + v^2) / 2
```

but instantiated types when specifying which architecture and advection scheme to use

```julia
model = IncompressibleModel(grid=grid, architecture=GPU(), advection=WENO5())
```

It seems there's some element of memorization to know which one to use when. Maybe this will confuse future users (certainly it confused us in #1118).

Changing this behavior would require a lot of refactoring but probably worth discussing whether changing makes sense or if we can make the user interface more robust and friendly to mismatches between data types and instantiated types?

### PR DESCRIPTION
**Pull request by navidcy**
This PR overhauls grid utils + adds `minimum_x/y/zspacing` methods that work on any grid.

(with @glwagner)

Closes #1119

### CODE REVIEW

File: src/Utils/cell_advection_timescale.jl [hunk 1]
```diff
@@ -7,9 +7,9 @@ function cell_advection_timescale(u, v, w, grid)
     vmax = maximum(abs, v)
     wmax = maximum(abs, w)
 
-    Δx = min_Δx(grid)
-    Δy = min_Δy(grid)
-    Δz = min_Δz(grid)
+    Δx = minimum_spacing(:x, (Face, Face, Face), grid)
+    Δy = minimum_spacing(:y, (Face, Face, Face), grid)
+    Δz = minimum_spacing(:z, (Face, Face, Face), grid)
```
Comment (navidcy) [inline src/Utils/cell_advection_timescale.jl:12]: should we be more accurate here using 
```suggestion
    Δx = minimum_spacing(:x, (Face, Center, Center), grid)
    Δy = minimum_spacing(:y, (Center, Face, Center), grid)
    Δz = minimum_spacing(:z, (Center, Center, Face), grid)
```
?

**Comment by navidcy** [comment]
@glwagner, I brought this commit of yours from #2867. 
This was my first cherry-pick commit; thanks @Angus-g!

**Review by glwagner — APPROVED** [review]

**Comment by navidcy** [comment]
@glwagner I think there is still an issue in regrid_xy, only on GPU now.

**Comment by navidcy** [comment]
omg, perhaps I fixed it... see [5272106](https://github.com/CliMA/Oceananigans.jl/pull/2991/commits/52721062f14c09da3abbe33c3de95a18397fb975)?

**Comment by navidcy** [comment]
now it's getting stuck on equality of two cuda subarrays... could you have a look at the GPU unit tests @glwagner?

File: src/Grids/grid_utils.jl [hunk 2]
```diff
@@ -41,7 +41,7 @@ function Base.:(==)(grid1::AbstractGrid, grid2::AbstractGrid)
     x1, y1, z1 = nodes(grid1, (Face(), Face(), Face()))
     x2, y2, z2 = nodes(grid2, (Face(), Face(), Face()))
 
-    return x1 == x2 && y1 == y2 && z1 == z2
+    CUDA.@allowscalar return x1 == x2 && y1 == y2 && z1 == z2
```
Comment (glwagner) [inline src/Grids/grid_utils.jl:44]: There's a nicer way to fix this --- this is sort of a hack that will be quite slow for large grids.

The nodes are using a view of an OffsetArray of a CuArray. From the stacktrace:

```
==(A::SubArray{Float64, 1, OffsetVector{Float64, CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}}, Tuple{UnitRange{Int64}}, true}, B::SubArray{Float64, 1, OffsetVector{Float64, CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}}, Tuple{UnitRange{Int64}}, true})
```

 There's no GPU-friendly `==` operator for that deeply nested object. However, there are GPU-friendly operators for subarray.

I think when we return `xnodes`, etc it's probably nicer to use a plain `view` around the underlying array.

We can leave this for this PR and I can open a new PR to fix it later.

Comment (navidcy) [reply src/Grids/grid_utils.jl:44]: Thanks. I was wondering yesterday about this.

**Comment by navidcy** [comment]
Seems likes docs are about to fail probably due to `interior(ω_timeseries, :, 1, :, 3)`. So perhaps we fix the grid equality on this PR?

**Comment by navidcy** [comment]
> Seems likes docs are about to fail probably due to `interior(ω_timeseries, :, 1, :, 3)`. So perhaps we fix the grid equality on this PR?

I fixed everything so we are back on the 6hr30min train of building the docs....

if you wanna fix the grid equality sometime soon then go for it, otherwise let's leave it for a future PR...

**Pull request merged**